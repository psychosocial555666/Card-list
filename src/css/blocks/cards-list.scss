.cards-list {
  padding: 0;
  margin: 0;
  list-style: none;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-column-gap: 50px;
  grid-row-gap: 50px;
}