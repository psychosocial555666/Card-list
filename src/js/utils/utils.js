
export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const checkTitleValidity = (value) => {
  return 4 < value.length && value.length <= 255
};

export const saveDataToStorage = (data, needToSave) => {
  let storage = window.localStorage;

  needToSave ? storage.setItem('cards', JSON.stringify(data)) : '';

  return data;
};

export const getDataFromStorage = () => {
  let storage = window.localStorage;

  return storage.cards ? JSON.parse(storage.cards) : [];
};

export const updateCardItem = (cards, card, data) => {
  let newCard = extend(card, data);
  let index = cards.findIndex((item) => {return item.id === card.id});

  let newCards = cards.slice(0, cards.length);
  newCards.splice(index, 1, newCard);

  return newCards;
};

export const deleteCardItem = (cards, card) => {
  let index = cards.findIndex((item) => {return item.id === card.id});

  let newCards = cards.slice(0, cards.length);
  newCards.splice(index, 1);

  return newCards;
};

export const pushNewCard = (cards, newCard) => {
  return cards.concat(newCard);
};

export const noop = () => {};

