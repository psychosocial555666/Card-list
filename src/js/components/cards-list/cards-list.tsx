import * as React from 'react';
import { getCardsList } from '../../reducer/data/selectors';
import { ActionCreator as ActionCreatorDATA } from "../../reducer/data/data";
import { connect } from "react-redux";
import { Card } from '../../types';
import edit from '../../../img/pencil-square.svg';
import plus from '../../../img/plus.svg';
import { Link } from 'react-router-dom';
import { extend } from '../../utils/utils';
let moment = require('moment');

interface Props {
  cards: Card[];
  updateCards: (cards: Card[], isNeedToSave: boolean) => void;
}

class CardsList extends React.PureComponent<Props, {}> {

  constructor(props: Props | Readonly<Props>) {
    super(props);
  }

  _likeChangeHandler(cards: Card[], i: number, status: boolean) {
    const {updateCards} = this.props;

    let newCard = extend(cards[i], {isLiked: !status});

    let newCards = cards.slice(0, cards.length);
    newCards.splice(i, 1, newCard);
    updateCards(newCards, true);
  }

  render() {
    const { cards } = this.props;

    return (
      <ul className="main-screen__list cards-list">
        <li className="cards-list__item card card--new">
          <Link to="/new">
            <img src={plus} alt="Add new card" width="70" height="70" />
          </Link>
        </li>
        {cards.length > 0 ? cards.map((item, i) => {
          return (
            <li key={i} className="cards-list__item card">
              <h2>{item.title}</h2>
              <p>{item.description}</p>
              <div className="card__controls">
                <Link to={`/edit/${item.id}`} 
                  className="card__edit"
                  >
                  <span className="visually-hidden">Edit</span>
                  <img src={edit} alt="Edit" width="30" height="30" />
                </Link>
                <div className="card__like">
                  <input type="checkbox" 
                  id={`like ${item.id}`} 
                  className="visually-hidden" 
                  checked={item.isLiked} 
                  onChange={() => {this._likeChangeHandler(cards, i, item.isLiked)}}
                  />
                  <label htmlFor={`like ${item.id}`}><span className="visually-hidden">Like</span></label>
                </div>
              </div>
              <span className="card__time">{moment(item.date).format(`hh:mm DD MMMM YYYY`)}</span>
            </li>
          )
        })
          : ''}
      </ul>
    );
  }
};

const mapStateToProps = (state: any) => ({
  cards: getCardsList(state),
});

const mapDispatchToProps = (dispatch: (arg0: any) => void) => ({
  updateCards(cards: any, isNeedToSave: boolean) {
    dispatch(ActionCreatorDATA.updateCards(cards, isNeedToSave));
  },
});

export { CardsList };

export default connect(mapStateToProps, mapDispatchToProps)(CardsList);