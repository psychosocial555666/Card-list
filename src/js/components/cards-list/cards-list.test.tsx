import * as renderer from "react-test-renderer";
import { CardsList } from "./cards-list";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { noop } from "../../utils/utils";
import { Router } from "react-router";
import history from "../../history";


const mockStore = configureStore([]);


const cards = [
  { id: '1vIfYnxecu4dIX_mB9Lzo', title: 'HTML и CSS, уровень 2', description: '2', isLiked: false, date: '2021-04-16T15:41:03.575Z' },
  { id: 'rOjKRPYaDR6LtrTfEWkMD', title: 'JavaScript, уровень 1', description: '3', isLiked: false, date: '2021-04-16T15:41:08.558Z' },
  { id: '8aBwJ6yriZdQusLFBMA56', title: 'Профессия Фронтэнд разработчик', description: '4', isLiked: false, date: '2021-04-16T15:41:14.214Z' },
  { id: 'm6T_3u7kf13vOPpmdQkTZ', title: 'JavaScript, уровень 2', description: '5', isLiked: false, date: '2021-04-16T15:41:25.294Z' }
]

describe(`CardsList`, () => {
  it(`Render CardsList`, () => {
    const store = mockStore({
      cards: cards,
      updateCards: () => { },
    });
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history}>
            <CardsList
              cards={cards}
              updateCards={noop}
            />
          </Router>

        </Provider>,
        {
          createNodeMock: () => {
            return document.createElement(`div`);
          }
        })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});