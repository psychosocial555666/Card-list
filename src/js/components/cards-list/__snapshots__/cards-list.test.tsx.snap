// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsList Render CardsList 1`] = `
<ul
  className="main-screen__list cards-list"
>
  <li
    className="cards-list__item card card--new"
  >
    <a
      href="/new"
      onClick={[Function]}
    >
      <img
        alt="Add new card"
        height="70"
        src="plus.svg"
        width="70"
      />
    </a>
  </li>
  <li
    className="cards-list__item card"
  >
    <h2>
      HTML и CSS, уровень 2
    </h2>
    <p>
      2
    </p>
    <div
      className="card__controls"
    >
      <a
        className="card__edit"
        href="/edit/1vIfYnxecu4dIX_mB9Lzo"
        onClick={[Function]}
      >
        <span
          className="visually-hidden"
        >
          Edit
        </span>
        <img
          alt="Edit"
          height="30"
          src="pencil-square.svg"
          width="30"
        />
      </a>
      <div
        className="card__like"
      >
        <input
          checked={false}
          className="visually-hidden"
          id="like 1vIfYnxecu4dIX_mB9Lzo"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="like 1vIfYnxecu4dIX_mB9Lzo"
        >
          <span
            className="visually-hidden"
          >
            Like
          </span>
        </label>
      </div>
    </div>
    <span
      className="card__time"
    >
      06:41 16 April 2021
    </span>
  </li>
  <li
    className="cards-list__item card"
  >
    <h2>
      JavaScript, уровень 1
    </h2>
    <p>
      3
    </p>
    <div
      className="card__controls"
    >
      <a
        className="card__edit"
        href="/edit/rOjKRPYaDR6LtrTfEWkMD"
        onClick={[Function]}
      >
        <span
          className="visually-hidden"
        >
          Edit
        </span>
        <img
          alt="Edit"
          height="30"
          src="pencil-square.svg"
          width="30"
        />
      </a>
      <div
        className="card__like"
      >
        <input
          checked={false}
          className="visually-hidden"
          id="like rOjKRPYaDR6LtrTfEWkMD"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="like rOjKRPYaDR6LtrTfEWkMD"
        >
          <span
            className="visually-hidden"
          >
            Like
          </span>
        </label>
      </div>
    </div>
    <span
      className="card__time"
    >
      06:41 16 April 2021
    </span>
  </li>
  <li
    className="cards-list__item card"
  >
    <h2>
      Профессия Фронтэнд разработчик
    </h2>
    <p>
      4
    </p>
    <div
      className="card__controls"
    >
      <a
        className="card__edit"
        href="/edit/8aBwJ6yriZdQusLFBMA56"
        onClick={[Function]}
      >
        <span
          className="visually-hidden"
        >
          Edit
        </span>
        <img
          alt="Edit"
          height="30"
          src="pencil-square.svg"
          width="30"
        />
      </a>
      <div
        className="card__like"
      >
        <input
          checked={false}
          className="visually-hidden"
          id="like 8aBwJ6yriZdQusLFBMA56"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="like 8aBwJ6yriZdQusLFBMA56"
        >
          <span
            className="visually-hidden"
          >
            Like
          </span>
        </label>
      </div>
    </div>
    <span
      className="card__time"
    >
      06:41 16 April 2021
    </span>
  </li>
  <li
    className="cards-list__item card"
  >
    <h2>
      JavaScript, уровень 2
    </h2>
    <p>
      5
    </p>
    <div
      className="card__controls"
    >
      <a
        className="card__edit"
        href="/edit/m6T_3u7kf13vOPpmdQkTZ"
        onClick={[Function]}
      >
        <span
          className="visually-hidden"
        >
          Edit
        </span>
        <img
          alt="Edit"
          height="30"
          src="pencil-square.svg"
          width="30"
        />
      </a>
      <div
        className="card__like"
      >
        <input
          checked={false}
          className="visually-hidden"
          id="like m6T_3u7kf13vOPpmdQkTZ"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="like m6T_3u7kf13vOPpmdQkTZ"
        >
          <span
            className="visually-hidden"
          >
            Like
          </span>
        </label>
      </div>
    </div>
    <span
      className="card__time"
    >
      06:41 16 April 2021
    </span>
  </li>
</ul>
`;
